Python/codecs.c

```diff
@@ -702,55 +702,67 @@ PyObject *PyCodec_IgnoreErrors(PyObject *exc)
 
 PyObject *PyCodec_ReplaceErrors(PyObject *exc)
 {
-    Py_ssize_t start, end, i, len;
+    Py_ssize_t start, end;
 
     if (PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_UnicodeEncodeError)) {
-        PyObject *res;
-        Py_UCS1 *outp;
-        if (PyUnicodeEncodeError_GetStart(exc, &start))
+        if (PyUnicodeEncodeError_GetStart(exc, &start) < 0) {
             return NULL;
-        if (PyUnicodeEncodeError_GetEnd(exc, &end))
+        }
+        if (PyUnicodeEncodeError_GetEnd(exc, &end) < 0) {
             return NULL;
-        len = end - start;
-        res = PyUnicode_New(len, '?');
-        if (res == NULL)
+        }
+        if (end <= start) {
+            goto oob;
+        }
+        Py_ssize_t len = end - start;
```

Comment:
This gets repeated (a lot, across the 3 PRs). Could you put it in a helper function?

Reply:
I can have a helper function but for each unicode error class since I need a different Get{Start,End} function.

Reply:
Would it make sense to add a `_PyUnicodeError_GetStartEndObject` helper that would give you all three, properly adjusted, and signal oob (`end <= start`) with the return value?

Reply:
AFAICS, what's needed from the object is the length: `PyObject_Size` should work for both `str` and `bytes`.
There's also type checking: IMO, consolidating `get_string` & `get_unicode` into one function that takes a flag could be an improvement -- there are only two places that are different, and the code duplication currently bubbles up into everything that uses them.
`PyBytes_Check` & `PyUnicode_Check` are macros, so function pointers wouldn't help here but you can expand them to get
`PyType_FastSubclass(Py_TYPE(op), is_bytes ? Py_TPFLAGS_BYTES_SUBCLASS : Py_TPFLAGS_UNICODE_SUBCLASS)`.

Reply:
I'll prepare a separate PR for refactoring. That way, I can update the 3 handler branches independently when needed.